### PR TITLE
CTP-3412 & CTP-3413 Create overrides for ECs & SORAs

### PR DESCRIPTION
**Pull request by aydevworks**
Deployed to 44-clc.
Not testable at the moment. It shouldn't break the plugin as the new code is not called anywhere.

### CODE REVIEW

File: classes/assessment/assessment.php [hunk 1]
```diff
@@ -196,6 +220,19 @@ public function check_assessment_validity(): \stdClass {
         return $result;
     }
 
+    /**
+     * Check if the assessment is valid for EC or SORA extension.
+     *
+     * @return bool
+     */
+    public function is_valid_for_extenstion(): \stdClass {
```
Comment (aspark21) [inline classes/assessment/assessment.php:228]: Typo in extension

Comment (leonstr) [reply classes/assessment/assessment.php:228]: Also, function declares return type `\stdClass`, PHPDoc says `@return bool`.

Comment (aydevworks) [reply classes/assessment/assessment.php]: Updated.

File: classes/assessment/assign.php [hunk 2]
```diff
@@ -53,4 +66,168 @@ public function get_start_date(): ?int {
...
         return $this->sourceinstance->duedate;
     }
+
+    /**
+     * Apply EC extension to the assessment.
+     *
+     * @param ec $ec The EC extension.
+     * @return void
+     */
+    protected function apply_ec_extension(ec $ec): void {
```
Comment (aspark21) [inline classes/assessment/assign.php:76]: just noting that if we need to implement this per activity, the other pull requests allowing push would only be providing partial support and would need follow up story to implement this in each

Comment (aydevworks) [reply classes/assessment/assign.php:76]: Yes, that's correct, and the `apply_sora_extension()` function too.

File: classes/settings/admin_setting_configdate.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ */
+
+namespace local_sitsgradepush\settings;
+
+use admin_setting_configtext;
+
+/**
+ * An admin setting for a date input.
+ */
+class admin_setting_configdate extends admin_setting_configtext {
```
Comment (aspark21) [inline classes/settings/admin_setting_configdate.php:33]: what is this for?

File: tests/extension/extension_test.php [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+        $this->resetAfterTest();
+
+        // Set admin user.
+        $this->setAdminUser();
+
+        // Get data generator.
+        $dg = $this->getDataGenerator();
+
+        // Setup testing environment.
+        set_config('late_summer_assessment_end_' . date('Y'), date('Y-m-d', strtotime('+2 month')), 'block_lifecycle');
```
Comment (aspark21) [inline tests/extension/extension_test.php:61]: https://moodledev.io/docs/4.4/apis/core/clock might be useful

**Review by aspark21 — APPROVED** [review]

File: classes/settings/admin_setting_configdate.php (hunk 3 above)
Comment (aydevworks) [reply classes/settings/admin_setting_configdate.php]: It implements a datepicker in a text config, but it's been removed now.
As discussed,  I added a field in the mapping table to distinguish which mappings could apply extensions.

File: tests/extension/extension_test.php (hunk 4 above)
Comment (aydevworks) [reply tests/extension/extension_test.php:61]: Thanks, but seems not applicable here.

**Comment by aydevworks** [comment]
Changes moved to [ec-sora-dev](https://github.com/ucl-isd/moodle-local_sitsgradepush/tree/ec-and-sora-dev) branch.
This PR is not needed.

**Pull request closed without merging**